File: core/paraphraser.py
import re
from typing import Dict, List, Any, Optional


class AcademicParaphraser:
    """
    Synthesizes and paraphrases overlapping sentences into distinct,
    scholarly formulations with proper attribution and syntax restructuring.
    """

    TRANSITIONS = [
        "Specifically, empirical investigations indicate that",
        "Notably, prior research demonstrates that",
        "Furthermore, contemporary findings substantiate that",
        "In this context, domain literature reveals that",
        "As rigorously observed in experimental trials,",
    ]

    VERBS = [
        ("demonstrates", "establishes"),
        ("shows", "substantiates"),
        ("proves", "validates"),
        ("analyzes", "evaluates"),
        ("predicts", "estimates"),
        ("explains", "elucidates"),
        ("uses", "utilizes"),
        ("finds", "uncovers"),
        ("helps", "facilitates"),
    ]

    @classmethod
    def synthesize_sentence(
        cls,
        sentence: str,
        source_name: Optional[str] = None,
        source_title: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Generates 3 distinct academic restructuring options:
        1. Active Inversion with Direct Attribution
        2. Passive Methodological Synthesis
        3. High-Density Conceptual Restructuring
        """
        clean_s = sentence.strip().rstrip(".!?")
        words = clean_s.split()
        if not words:
            return {"suggestions": []}

        clean_author = "the cited literature"
        if source_title:
            clean_author = f"{source_title}"
        elif source_name:
            clean_author = source_name.replace(".txt", "").replace(".pdf", "").replace("_", " ").title()

        # Option 1: Active Inversion with Authorial Attribution
        first_word_lower = words[0].lower() + " " + " ".join(words[1:])
        opt1 = f"According to {clean_author}, {first_word_lower}."

        # Option 2: Methodological Synthesis with Inversion
        syn_verb_applied = first_word_lower
        for old_v, new_v in cls.VERBS:
            if f" {old_v} " in syn_verb_applied:
                syn_verb_applied = syn_verb_applied.replace(f" {old_v} ", f" {new_v} ", 1)
                break
        
        opt2 = f"Empirical findings suggest that {syn_verb_applied} ({clean_author}, 2024)."

        # Option 3: High-Density Analytical Rephrasing
        if len(words) > 6:
            mid = len(words) // 2
            first_half = " ".join(words[:mid])
            second_half = " ".join(words[mid:])
            opt3 = f"Regarding {first_half.lower()}, subsequent analysis confirms that {second_half}."
        else:
            opt3 = f"Notably, {first_word_lower}, reflecting core principles documented in {clean_author}."

        return {
            "original_sentence": sentence,
            "source_attributed": clean_author,
            "suggestions": [
                {
                    "style": "Authorial Attribution (Active)",
                    "text": opt1,
                    "focus": "Direct scholarly attribution with formal introductory clause."
                },
                {
                    "style": "Methodological Synthesis (Passive)",
                    "text": opt2,
                    "focus": "Passive academic synthesis with parenthetical reference."
                },
                {
                    "style": "Analytical Restructuring (High-Density)",
                    "text": opt3,
                    "focus": "Sentence structure inversion with emphasis on findings."
                }
            ]
        }

File: lib/paraphraser.js
/**
 * Academic Paraphraser & Synthesis Engine
 * Provides 1-click academic restructuring and formal attribution formulations.
 */

const TRANSITIONS = [
    "Specifically, empirical investigations indicate that",
    "Notably, prior research demonstrates that",
    "Furthermore, contemporary findings substantiate that",
    "In this context, domain literature reveals that",
    "As rigorously observed in experimental trials,"
];

const VERBS = [
    [/\bdemonstrates\b/gi, "establishes"],
    [/\bshows\b/gi, "substantiates"],
    [/\bproves\b/gi, "validates"],
    [/\banalyzes\b/gi, "evaluates"],
    [/\bpredicts\b/gi, "estimates"],
    [/\bexplains\b/gi, "elucidates"],
    [/\buses\b/gi, "utilizes"],
    [/\bfinds\b/gi, "uncovers"],
    [/\bhelps\b/gi, "facilitates"]
];

class AcademicParaphraser {
    /**
     * Synthesizes and paraphrases an overlapping sentence into 3 distinct academic options.
     * @param {string} sentence 
     * @param {string} [sourceName] 
     * @param {string} [sourceTitle] 
     */
    static synthesizeSentence(sentence, sourceName = null, sourceTitle = null) {
        const cleanS = (sentence || "").trim().replace(/[.!?]+$/, "");
        if (!cleanS) {
            return { original_sentence: "", suggestions: [] };
        }

        const words = cleanS.split(/\s+/);
        const citeKey = (sourceTitle || sourceName || "Cited Publication").replace(/\.[^/.]+$/, "");
        const transition = TRANSITIONS[Math.abs(cleanS.length) % TRANSITIONS.length];

        let restructuredV1 = cleanS;
        for (const [pattern, replacement] of VERBS) {
            if (pattern.test(restructuredV1)) {
                restructuredV1 = restructuredV1.replace(pattern, replacement);
                break;
            }
        }

        // 1. Active Inversion with Direct Attribution
        const opt1 = `According to ${citeKey}, ${restructuredV1.charAt(0).toLowerCase() + restructuredV1.slice(1)}.`;

        // 2. Passive Methodological Synthesis
        const opt2 = `${transition} ${cleanS.charAt(0).toLowerCase() + cleanS.slice(1)} (${citeKey}, 2024).`;

        // 3. High-Density Conceptual Restructuring
        let opt3 = "";
        if (words.length > 7) {
            const splitIdx = Math.floor(words.length / 2);
            const firstHalf = words.slice(0, splitIdx).join(" ");
            const secondHalf = words.slice(splitIdx).join(" ");
            opt3 = `Regarding ${secondHalf.toLowerCase()}, ${firstHalf.charAt(0).toLowerCase() + firstHalf.slice(1)} as established in recent literature [${citeKey}].`;
        } else {
            opt3 = `Evidence presented in ${citeKey} reinforces that ${cleanS.toLowerCase()}.`;
        }

        return {
            original_sentence: sentence,
            source_attribution: citeKey,
            suggestions: [
                {
                    style: "Active Attribution (APA/MLA)",
                    text: opt1,
                    description: "Explicit attribution clause framing the author's primary assertion."
                },
                {
                    style: "Methodological Synthesis (Scholarly Formal)",
                    text: opt2,
                    description: "Passive voice synthesis contextualizing findings into academic consensus."
                },
                {
                    style: "Conceptual Restructuring (High-Density)",
                    text: opt3,
                    description: "Inverted focus prioritizing empirical implications over source phrasing."
                }
            ]
        };
    }
}

module.exports = { AcademicParaphraser };
